tests/unit/test_routing_manager.cpp

```diff
@@ -3118,6 +3126,279 @@ void TestNat64PrefixSelection(void)
+
+    VerifyOrQuit(otBorderRoutingGetPdOmrPrefix(sInstance, &platformPrefixInfo) == OT_ERROR_NONE);
+    VerifyOrQuit(AsCoreType(&platformPrefixInfo.mPrefix) == aPrefix);
+}
+
+void VerifyNoPdOmrPrefix()
+{
+    otBorderRoutingPrefixTableEntry platformPrefixInfo;
+
+    VerifyOrQuit(otBorderRoutingGetPdOmrPrefix(sInstance, &platformPrefixInfo) == OT_ERROR_NOT_FOUND);
+}
+
+void TestBorderRoutingProcessPlatfromGeneratedNd(void)
+{
+    Ip6::Prefix localOmr;
+
+    Log("--------------------------------------------------------------------------------------------");
+    Log("TestBorderRoutingProcessPlatfromGeneratedNd");
+
+    InitTest(/* aEnableBorderRouting */ true);
+
+    otBorderRoutingDhcp6PdSetEnabled(sInstance, true);
+
+    {
+        SuccessOrQuit(sInstance->Get<BorderRouter::RoutingManager>().GetOmrPrefix(localOmr));
+    }
+
+    // 0. Reject invalid RA.
+    Log("0. Invalid RA message.");
+    {
+        {
+            const uint8_t testInvalidRaMessage[] = {
+                0x86, 0x00, 0x00, 0x00, 0x40, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            };
+
+            otPlatBorderRoutingProcessIcmp6Ra(sInstance, testInvalidRaMessage, sizeof(testInvalidRaMessage));
+            VerifyNoPdOmrPrefix();
+        }
+
+        {
+            const uint8_t testInvalidRaMessage[] = {
+                0x87, 0x00, 0x00, 0x00, 0x40, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            };
+
+            otPlatBorderRoutingProcessIcmp6Ra(sInstance, testInvalidRaMessage, sizeof(testInvalidRaMessage));
+            VerifyNoPdOmrPrefix();
+        }
+
+        {
+            const uint8_t testRaMessageWithInvalidPrefix[] = {
+                0x86, 0x00, 0x00, 0x00, 0x40, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x03, 0x04, 0x41, 0xc0, 0x00, 0x00, 0x10, 0xe1, 0x00, 0x00, 0x04, 0xd2, 0x00, 0x00, 0x00, 0x00,
+                0x20, 0x01, 0x0d, 0xb8, 0x00, 0x01, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            };
+
+            otPlatBorderRoutingProcessIcmp6Ra(sInstance, testRaMessageWithInvalidPrefix,
+                                              sizeof(testRaMessageWithInvalidPrefix));
+            VerifyNoPdOmrPrefix();
+        }
+    }
+
+    // 1. Publish a prefix, and wait until it expired.
+    Log("1. Simple RA message.");
+    {
+        Ip6::Prefix raPrefix = PrefixFromString("2001:db8:dead:beef::", 64);
+        // A simple RA message
+        // PIO: 2001:db8:dead:beef::/64, Pref 1234s, Valid 4321s
+        // Note: Checksum is ignored.
+        const uint8_t testValidRaMessage[] = {
+            0x86, 0x00, 0x00, 0x00, 0x40, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0x03, 0x04, 0x40, 0xc0, 0x00, 0x00, 0x10, 0xe1, 0x00, 0x00, 0x04, 0xd2, 0x00, 0x00, 0x00, 0x00,
+            0x20, 0x01, 0x0d, 0xb8, 0xde, 0xad, 0xbe, 0xef, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        };
```

Comment:
This works ok, but we can also try to construct the RA messages (using OT methods for `Ip6::Nd::RouterAdvertMessage`)?

Reply:
We have this function (and its template variants) which prepares a RA message with a bunch of PIO and RIO entries and send it (pass it to `otPlatInfraIfRecvIcmp6Nd()` to be processed by OT core), we can refactor this and seperate out the code where we prepare the RA message (into its own helper function) so we can use it in this test for DHCP6 PD (and prepare the RA).
```c++
void SendRouterAdvert(const Ip6::Address &aRouterAddress,
                      const Pio          *aPios,
                      uint16_t            aNumPios,
                      const Rio          *aRios,
                      uint16_t            aNumRios,
                      const DefaultRoute &aDefaultRoute);

template <uint16_t kNumPios, uint16_t kNumRios>
void SendRouterAdvert(const Ip6::Address &aRouterAddress,
                      const Pio (&aPios)[kNumPios],
                      const Rio (&aRios)[kNumRios],
                      const DefaultRoute &aDefaultRoute = DefaultRoute(0, NetworkData::kRoutePreferenceMedium));

template <uint16_t kNumPios>
void SendRouterAdvert(const Ip6::Address &aRouterAddress,
                      const Pio (&aPios)[kNumPios],
                      const DefaultRoute &aDefaultRoute = DefaultRoute(0, NetworkData::kRoutePreferenceMedium));

.```

Reply:
Did a small refactor for BuildRouterAdvert here for smaller changeset.

Suggest another refactor to decouple `otPlatInfraIfRecvIcmp6Nd` from `SendRouterAdvert`

Reply:
Thanks. 👍